### PR TITLE
Handle non-printable Go test result gracefully

### LINKED ISSUE #2113: Buck crash when reading Go test output with invalid characters
Go test output can contains invalid characters because of, for example, invalid user inputs. The native "go test" output handles them nicely, but Buck would crash.

Steps to reproduce:
```shell
$ git clone git@github.com:linzhp/buck_go_examples.git
$ cd buck_go_examples
$ go test ./testOutput
--- FAIL: TestUnprintableChars (0.00s)
    testOutput_test.go:13: ����JFIF is not printable
FAIL
FAIL	buck_go_examples/testOutput	0.006s
$ buck test testOutput:
Action graph will be rebuilt because files have been added or removed.
FAILURE //testOutput:all_tests main: Input length = 1
Building: finished in 0.1 sec (100%) 2/2 jobs, 0 updated
  Total time: 0.2 sec
Testing: finished in 0.0 sec (0 PASS/1 FAIL)
RESULTS FOR //testOutput:all_tests
FAIL    <100ms  0 Passed   0 Skipped   1 Failed   //testOutput:all_tests
FAILURE //testOutput:all_tests main: Input length = 1
java.nio.charset.MalformedInputException: Input length = 1
	at java.nio.charset.CoderResult.throwException(CoderResult.java:281)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:339)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at com.facebook.buck.features.go.GoTest.parseTestResults(GoTest.java:168)
	at com.facebook.buck.features.go.GoTest.lambda$interpretTestResults$0(GoTest.java:251)
	at com.facebook.buck.cli.TestRunning$4.call(TestRunning.java:547)
	at com.facebook.buck.cli.TestRunning$4.call(TestRunning.java:540)
	at com.facebook.buck.cli.TestRunning.lambda$runStepsAndYieldResult$1(TestRunning.java:824)
	at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$submitWithSemaphore$0(WeightedListeningExecutorService.java:74)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:206)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:195)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:115)
	at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:999)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

TESTS FAILED: 1 FAILURE
Failed target: //testOutput:all_tests
FAIL //testOutput:all_tests
```

@mkaczanowski

### PR DESCRIPTION
**Pull request by linzhp**
Make Buck replace non-printable characters in Go test result with decoder's default value, and continue parsing the result, thus avoid crashing with uncatched exception. Fixed #2113 

@bobyangyf Please review, thanks!

### CODE REVIEW

**Comment by linzhp** [comment]
@styurin @ttsugriy @philipjameson  Could you take a look?

**Comment by ttsugriy** [comment]
sorry @linzhp, I no longer work on Buck, but @styurin or @philipjameson should be able to help.

File: src/com/facebook/buck/features/go/GoTest.java [hunk 1]
```diff
@@ -159,9 +162,14 @@ public GoTest(
 
   private ImmutableList<TestResultSummary> parseTestResults() throws IOException {
     ImmutableList.Builder<TestResultSummary> summariesBuilder = ImmutableList.builder();
+    CharsetDecoder decoder = Charsets.UTF_8.newDecoder();
+    decoder.onMalformedInput(CodingErrorAction.REPLACE);
+    decoder.replaceWith("囧");
```
Comment (bobyangyf) [inline src/com/facebook/buck/features/go/GoTest.java]: whats with this character?

Comment (linzhp) [reply src/com/facebook/buck/features/go/GoTest.java]: Something to use as the placeholder for non-printable characters. If you don't like this one, please suggest another one, I will replace.

Comment (bobyangyf) [reply src/com/facebook/buck/features/go/GoTest.java]: lets make this atleast a named constant

File: test/com/facebook/buck/features/go/testdata/go_test/testOutput/BUCK.fixture [hunk 2]
```diff
@@ -0,0 +1,5 @@
+go_test(
+  name='all_tests',
```
Comment (LLITCHEV) [inline test/com/facebook/buck/features/go/testdata/go_test/testOutput/BUCK.fixture]: Could you indent these lines with 4 spaces. Our formatter complains and will prevent checkin in our internal repo.

**Comment by LLITCHEV** [comment]
Hi, sorry for the hustle...
Our internal linter is still giving warning - we use buildifier internally...
Here is the exact warning for file test/com/facebook/buck/features/go/testdata/go_test/testOutput/
BUCK.fixture

go_test(
    name='all_tests',
    srcs=glob(['*_test.go']),
    resources=["file.jpg"]
)
phabricatorlintersuggested changes to lines 1-5
BUCK files should be formatted using buildifier.
See https://fburl.com/auto-format-build-files for details.
Lint code: BUILDIFIERFORMAT
Lint name: buildifier formatting violation
 go_test(
-    name='all_tests',
-    srcs=glob(['*_test.go']),
-    resources=["file.jpg"]
+    name = "all_tests",
+    srcs = glob(["*_test.go"]),
+    resources = ["file.jpg"],
 )

**Comment by styurin** [comment]
Can you use some simple file that contains small amount of binary data instead of a jpeg file?

**Comment by linzhp** [comment]
reduced the binary file size and rebased onto master. @styurin

File: src/com/facebook/buck/features/go/GoTest.java [hunk 3]
```diff
@@ -69,12 +72,14 @@
...
 public class GoTest extends AbstractBuildRuleWithDeclaredAndExtraDeps
     implements TestRule, HasRuntimeDeps, ExternalTestRunnerRule, BinaryBuildRule {
+
   private static final Pattern TEST_START_PATTERN = Pattern.compile("^=== RUN\\s+(?<name>.*)$");
   private static final Pattern TEST_FINISHED_PATTERN =
       Pattern.compile(
           "^\\s*--- (?<status>PASS|FAIL|SKIP): (?<name>.+) \\((?<duration>\\d+\\.\\d+)(?: seconds|s)\\)$");
   // Extra time to wait for the process to exit on top of the test timeout
   private static final int PROCESS_TIMEOUT_EXTRA_MS = 5000;
+  private static final String NON_PRINTABLE_REPLACEMENT = "囧";
```
Comment (styurin) [inline src/com/facebook/buck/features/go/GoTest.java]: What is this character?

Comment (styurin) [reply src/com/facebook/buck/features/go/GoTest.java]: Is that something that `go test` print? Why not use some ASCII character?

**Comment by styurin** [comment]
Some internal feedback:

> This is better than the jpeg and is fine, but really we don't even need a file do we? All they do with it is read it in the go test to get binary data that is unprintable. Is there no way to create that binary data in the test itself? Then you could make it clear what specific values are being tested (and if, like Java, there are actual invalid sequences you could test those too).

**Comment by linzhp** [comment]
@styurin That's a good point. I've updated the diff to avoid reading files all together.

**Pull request closed without merging**